### PR TITLE
VR+tgui additions:

### PR DESCRIPTION
**Pull request by RemieRichards**
:cl:
add: Added VR Sleepers that allow users to enter a virtual body in a virtual world (Currently not mapped in ;_;)
/:cl:
## VR
- Adds "VR Sleepers" which allow the user to enter a virtual body in a virtual world (it's basically a glorified teleporter)
- There are currently no vr spawn points set on the map, so the vr sleepers do not work (the vr sleepers are also not mapped, I look to our mappers to make vr areas/map in vr sleepers)
- If you do not have a virtual body, one will be created as a copy of you (provided the sleeper allows it, there's a setting to disable it incase a future vr game/scenario needs you to play a certain role)
- When you leave VR your virtual body is NOT deleted, meaning whomever gets into that vr sleeper can steal your virtual body, to prevent this you must manually delete your avatar using the UI 
- Emagging a vr sleeper means that if the user of that vr sleeper dies in VR, they die for real!
## TGUI
- Allows tgui buttons to process fontawesome icons in stacks, fontawesome icons can stack over each other to fake a new icon (eg a ban icon over a camera to mean "ban cameras")
- tgui buttons are now a MINIMUM of 20px high, as opposed to being permanently 20px high, this is so that large fontawesome icons don't look hideous.
- like how tgui attribute `icon`'s syntax is a simplified version of fontawesome's own syntax, `icon_stack` is also simplified: `square-o 2x, twitter 1x` is automatically converted to valid fontawesome+HTML (image 3)
- `icon_stack` is parsed using commas and regex, you must specify icons in the form `iconname sizemultiplier` placing a comma between each: `iconname sizemultiplier, iconname2 sizemultiplier2`, the regex used to parse the icon and its size allows for whitespace: `square-o               2x` and the lack of it: `square-o2x,twitter1x`, but you MUST use commas between each icon.
  @neersighted if you're there, mind looking at my JS? it's functional but you know, standards.
## Images

<B>Image 1</B> - VRception (ignore clothes changing, that's a bug that was fixed)
vrception
<B>Image 2</B>\- VR sleeper's UI
vr-sleeper-tgui
<B>Image 3</B> - Icon Stacks working in tgui (there are 4 because I was testing)
twitter-square-icon-stack

### CODE REVIEW

File: code/modules/VR/vr_sleeper.dm [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+//Glorified teleporter that puts you in a new human body.
+// it's """VR"""
+/obj/machinery/vr_sleeper
+	name = "virtual reality sleeper"
+	desc = "a sleeper modified to alter the subconscious state of the user, allowing them to visit virtual worlds"
+	icon = 'icons/obj/Cryogenic2.dmi'
+	icon_state = "sleeper"
+	state_open = TRUE
+	anchored = TRUE
+	var/you_die_in_the_game_you_die_for_real = FALSE
```
Comment (Fox-McCloud) [inline code/modules/VR/vr_sleeper.dm]: gosh damn it, @RemieRichards

File: code/modules/VR/vr_sleeper.dm [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+	return safepick(available_vr_spawnpoints[vr_category])
+
+
+/obj/machinery/vr_sleeper/proc/build_virtual_human(mob/living/carbon/human/H, location, transfer = TRUE)
+	if(H)
+		cleanup_vr_human()
+		vr_human = new /mob/living/carbon/human/virtual_reality(location)
+		vr_human.vr_sleeper = src
+		vr_human.real_me = H
+		H.dna.transfer_identity(vr_human)
```
Comment (Fox-McCloud) [inline code/modules/VR/vr_sleeper.dm]: Does this handle species changes as well? If not, you may want to add that

File: code/modules/VR/vr_human.dm [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+/mob/living/carbon/human/virtual_reality/New()
+	..()
+	quit_action = new()
+	quit_action.Grant(src)
+
+
+/mob/living/carbon/human/virtual_reality/death()
+	revert_to_reality()
+	..()
```
Comment (Fox-McCloud) [inline code/modules/VR/vr_human.dm]: I'd recommend deleting the mob, lest we have a VR room with 50 dead bodies from someone entering and suiciding a billion times.

**Comment by Fox-McCloud** [comment]
>When you leave VR your virtual body is NOT deleted, meaning whomever gets into that vr sleeper can steal your virtual body, to prevent this you must manually delete your avatar using the UI 

Kinda weird, but ok, I suppose?

:+1: to this feature. I always liked Goon's VR--it was a great place to go to do whatever, without consequence and not have to worry about things.

Want to go to VR land and beat the shit out of everyone 4noraisin? That's ok. Much better combat situations and scenarios too other than just on the holodeck or with toy laser guns.

**Comment by xxalpha** [comment]
The age of cyberdecking approaches.

File: code/modules/VR/vr_sleeper.dm (hunk 2 above)
Comment (Fox-McCloud) [inline code/modules/VR/vr_sleeper.dm]: answered my own question---yes---yes it does.

**Comment by Jalleo** [comment]
This I am fine with if it has a joke added to end of nuke round of everyone waking up in a VR sleeper.

**Comment by RemieRichards** [comment]
@Jalleo that... that is fucking genius.

**Comment by Fox-McCloud** [comment]
@RemieRichards 

Might I suggest an option for duplicating the VR individual's inventory? Or at least just basic equiped clothes or something.

**Comment by RemieRichards** [comment]
@Fox-McCloud Yes that was something I forgot, I was just going to have them default to like, a random coloured jumpsuit or something (as a configurable outfit datum on the sleeper)

**Comment by Jalleo** [comment]
I know and that is why the joke should be done.
This is a great feature anyhow.

**Comment by Jazaen** [comment]
Hm, are VR races the same as what entered the simulation? Could you maybe change them to synths?

**Comment by KorPhaeron** [comment]
If you're gonna do "the nuke was all a dream" please do so in another PR because I'm sure that will cause a giant lore fight and be controversial and I'd rather this PR not be held up because of it

**Comment by Shadowlight213** [comment]
I kept wanting to do this myself, but never got around to it. :+1:
If you don't plan on doing it, I'll look into making vr goggles and put them in perma. However I'm iffy on not deleting the vr body when the user leaves, as it makes vr goggles much more unintuitive, and doesn't really seem like it's really that useful. A better way would be allowing the vr sleeper to be hacked and an alternate appearance be loaded from the manifest somehow.

**Comment by ike709** [comment]
do the nuke vr thing (in another PR)

**Comment by Anonmare** [comment]
Emagged sleepers shouldn't have an option to leave tbh.

**Comment by peoplearestrange** [comment]
Do we have any examples of what the VR world looks like?
Not really playing on goon this is all kinda new to me.

Would it be possible to make the VR world look like more shitty? Or like make everyone a full colour AI hologram (basically flicker and have there alpha down a little). Something that would make it cheesy and not just a clone of what you already are playing.

**Comment by Fox-McCloud** [comment]
>Do we have any examples of what the VR world looks like?

It's a very silly place. It's essentially a race track with the their version of the thunderdome in the middle of it all. Detective has his own private office on a beach, inaccessible by main VR attendees.

Their toxin bomb testing range is also in the VR, in a different section.

**Comment by hornygranny** [comment]
please make VR use the horrible horrible oldstation sprites

**Pull request merged**